libraries/chain/sps_objects/sps_processor.cpp

```diff
@@ -0,0 +1,286 @@
+{
+   uint64_t ret = 0;
+
+   const auto& pvidx = db.get_index< proposal_vote_index >().indices().get< by_proposal_voter >();
+   auto found = pvidx.find( id );
+
+   while( found != pvidx.end() && found->proposal_id == id )
+   {
+      const auto& _voter = db.get_account( found->voter );
+
+      //If _voter has set proxy, then his votes aren't taken into consideration
+      if( _voter.proxy == STEEM_PROXY_TO_SELF_ACCOUNT )
+      {
+         auto sum = std::accumulate(   _voter.proxied_vsf_votes.begin(),
+                                       _voter.proxied_vsf_votes.end(),
+                                       _voter.vesting_shares.amount );
+         ret += sum.value;
+      }
+
+      ++found;
+   }
+
+   return ret;
+}
+
+void sps_processor::calculate_votes( const t_proposals& proposals )
+{
+   for( auto& item : proposals )
+   {
+      const proposal_object& _item = item;
+      auto total_votes = calculate_votes( _item.proposal_id );
+
+      db.modify( _item, [&]( auto& proposal )
+                        {
+                           proposal.total_votes = total_votes;
+                        } );
+   }
+}
+
+void sps_processor::sort_by_votes( t_proposals& proposals )
+{ 
+   std::sort( proposals.begin(), proposals.end(), []( const proposal_object& a, const proposal_object& b )
+                                                      {
+                                                         if (a.total_votes == b.total_votes)
+                                                         {
+                                                            return a.proposal_id < b.proposal_id;
+                                                         }
+                                                         return a.total_votes > b.total_votes;
+                                                      } );
+}
+
+asset sps_processor::get_treasury_fund()
+{
+   auto& treasury_account = db.get_account( STEEM_TREASURY_ACCOUNT );
+
+   return treasury_account.sbd_balance;
+}
+
+asset sps_processor::get_daily_inflation()
+{
+   //temporary
+   return asset( 0, SBD_SYMBOL );
+}
+
+asset sps_processor::calculate_maintenance_budget( const time_point_sec& head_time )
+{
+   //Get funds from 'treasury' account ( treasury_fund ) 
+   asset treasury_fund = get_treasury_fund();
+
+   //Get daily proposal inflation ( daily_proposal_inflation )
+   asset daily_inflation = get_daily_inflation();
+
+   FC_ASSERT( treasury_fund.symbol == daily_inflation.symbol, "symbols must be the same" );
+
+   //Calculate budget for given maintenance period
+   auto passed_time_seconds = ( head_time - db.get_dynamic_global_properties().last_budget_time ).to_seconds();
+
+   //Calculate daily_budget_limit
+   auto daily_budget_limit = treasury_fund.amount.value / 100 + daily_inflation.amount.value;
+   daily_budget_limit = daily_budget_limit * ( static_cast< double >( passed_time_seconds ) / daily_seconds );
```

Comment:
casting to a double is a *very* bad idea. Floating point math is notoriously non-deterministic which could lead to a fork in the chain. Please replace all numerical declarations of `auto` with the correct integer type to ensure we do not release non-deterministic code.

Reply:
To get this as a fraction, we use `STEEM_1_PERCENT` and `STEEM_100_PERCENT` constants.

This should be rewritten as:

```
daily_budget_limit = ( daily_budget_limit * passed_time_seconds ) / ( daily_seconds * STEEM_100_PERCENT );
```

Developer analysis needs to be done to ensure that 64 bit fields cannot overflow in all cases. To be safe, it should probably be cast to fc's 128 bit integer type because assets can be up to 64 bits and a timestamp can be up to 32 bits. The denominator needs to be cast to a 64 bit integer to accommodate the multiplications between 32 bit and 16 bit integers.